configuration/config2/XMLEnvironmentLoader.cpp

```diff
@@ -0,0 +1,126 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2018 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "XMLEnvironmentLoader.hpp"
+#include "XSDSchemaParser.hpp"
+#include "Exceptions.hpp"
+
+
+std::vector<std::shared_ptr<EnvironmentNode>> XMLEnvironmentLoader::load(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem) const
+{
+    std::vector<std::shared_ptr<EnvironmentNode>> envNodes;
+    std::shared_ptr<EnvironmentNode> pEnvNode;
+
+    pt::ptree envTree;
+    try
+    {
+        pt::read_xml(in, envTree, pt::xml_parser::trim_whitespace | pt::xml_parser::no_comments);
+        //auto rootIt = envTree.begin();
+
+        // if root, want to start with rootIt, but auto match rootIt first with schemaItem, then parse on down, but want to return
+        // envNode for the root
+
+        // if not root, to through rootIt with schema item as parent for each elemement, return envNode of each item iterated
+
+        std::shared_ptr<SchemaItem> pParseRootSchemaItem;
+        for (auto envIt = envTree.begin(); envIt != envTree.end(); ++envIt)
+        {
+            if (envIt->first == pSchemaItem->getProperty("name"))
+            {
+                pParseRootSchemaItem = pSchemaItem;
+            }
+            else
+            {
+                pParseRootSchemaItem = pSchemaItem->getChild(envIt->first);
+            }
+
+            pEnvNode = std::make_shared<EnvironmentNode>(pParseRootSchemaItem, envIt->first);  // caller may need to set the parent
+            parse(envIt->second, pParseRootSchemaItem, pEnvNode);
+            envNodes.push_back(pEnvNode);
+        }
+    }
+    catch (const std::exception &e)
+    {
+        std::string xmlError = e.what();
+        std::string msg = "Unable to read/parse Environment file. Error = " + xmlError;
+        throw (ParseException(msg));
+    }
+    return envNodes;
+}
+
+
+void XMLEnvironmentLoader::parse(const pt::ptree &envTree, const std::shared_ptr<SchemaItem> &pConfigItem, std::shared_ptr<EnvironmentNode> &pEnvNode) const
```

Comment:
didn't notice any member vars, is this method static by nature/should it be declared as such

Reply:
The intent is that you use this method when you have an instance of the object, not to be able to call it otherwise. In the future there could be use of class member variables.